crates/matrix-sdk-crypto/src/ciphers.rs

```diff
@@ -0,0 +1,239 @@
+// Copyright 2023 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use aes::{
+    cipher::{generic_array::GenericArray, IvSizeUser, KeyIvInit, KeySizeUser, StreamCipher},
+    Aes256,
+};
+use ctr::Ctr128BE;
+use hmac::{
+    digest::{FixedOutput, MacError},
+    Hmac, Mac as _,
+};
+use pbkdf2::pbkdf2;
+use rand::{thread_rng, RngCore};
+use sha2::{Sha256, Sha512};
+use zeroize::{Zeroize, ZeroizeOnDrop};
+
+// We could use the `keysize()` method Aes256Ctr as KeySize exposes, but it's
+// not const (yet?), same for the IV size.
+pub(crate) const IV_SIZE: usize = 16;
+pub(crate) const KEY_SIZE: usize = 32;
+pub(crate) const SALT_SIZE: usize = 16;
+pub(crate) const MAC_SIZE: usize = 32;
+
+type Aes256Ctr = Ctr128BE<Aes256>;
+
+type Aes256Key = GenericArray<u8, <Aes256 as KeySizeUser>::KeySize>;
+type Aes256Iv = GenericArray<u8, <Aes256Ctr as IvSizeUser>::IvSize>;
+type HmacSha256Key = [u8; KEY_SIZE];
```

Comment:
Being boxes how? Which exact type do you have in mind for `Aes256IV`?

We could define `HmacSha256Key` as:

```rust
type HmacSha256Key<'a> = &'a [u8; KEY_SIZE];
```

Not sure if it makes any difference though.

Reply:
Sorry, `Aes256IV` doesn't need to be boxed since IVs generally don't need to be secret anyway.

The types I had in mind were `Box<GenericArray<u8, <Aes256 as KeySizeUser>::KeySize>>` and `Box<[u8; KEY_SIZE]>`.

> ```rust
> type HmacSha256Key<'a> = &'a [u8; KEY_SIZE];
> ```

That works just as well. It's not extremely important, but the point was to make it harder for someone else refactoring it to use the type in an improper way.

Reply:
If it's too fiddly, it's fine to leave it as it is btw.

Reply:
I think changing things for `GenericArray` is a nightmare. Getting every method to agree that things are now boxed might not even be possible.